chore:slack kit example app readme (#76)
knocklabs/javascript: examples/slack-kit-example/README.md

SlackKit is a collection of embeddable UI components and APIs that radically simplifies the process of adding a Slack integration to a SaaS product. This is an example app meant to help you get started using the SlackKit components and APIs provided by Knock.

## Getting started

For this project, there are quite a few configuration variables you need before you can get started. Some of these are typical environment variables you would supply to your application, like API keys, tokens, or client ids.

The other set of values would typically be determined by your product's business logic, but you can hardcode these values for the time being.

> ### Setting up a Slack app

Author: Yeah, agreed. The DRYer we can make this the better

| NEXT_PUBLIC_REDIRECT_URL           | This value comes from your application. It is where Knock will redirect your user after the OAuth flow with Slack. The default of `http://localhost:3000` is valid when running this project locally.                                                                                         |

### Knock resource variables

To make the connection between Slack channels and Knock objects, you'll also need to provide details for several types of resources in Knock. To do this, you can replace the values in the `getAppDetails` function inside of the `/app/lib/app-details.ts` file. These values would typically be determined by your application's business logic, but we can hardcode them for this example:

```
export function getAppDetails() {
  return {
    tenant: "knock-projects",
    collection: "repositories",
    objectId: "repo-2",
    userId: "123",
    workflowKey: "new-issue",
  };
}
```

You should already have a value for `workflowKey` from a previous step, and you can choose a [user identifier](https://docs.knock.app/concepts/users#user-identifiers) to use from the dashboard.

#### Create a tenant

> In Knock, [tenants](https://docs.knock.app/concepts/tenants) are an important concept, and with SlackKit they are used to store the access token for an organization's Slack workspace. You can create a new tenant from the dashboard and include it's ID as the value for the `tenant` property in the `getAppDetails` function. You can also use this cURL command to create a tenant by replacing the values for `tenant-id`, `KNOCK_API_KEY`, and `tenant-name`:

Reviewer: ```suggestion
In SlackKit, [tenants](https://docs.knock.app/concepts/tenants) are used to store the access token for an organization's Slack workspace. You can create a new tenant from the dashboard and include it's ID as the value for the `tenant` property in the `getAppDetails` function. You can also use this cURL command to create a tenant by replacing the values for `tenant-id`, `KNOCK_API_KEY`, and `tenant-name`:
```

Reviewer: ii think if we're not going to elaborate let's leave out that they're important here

```
curl --location --request PUT 'https://api.knock.app/v1/tenants/<tenant-id>' \
--header 'Content-Type: application/json' \
--header 'Authorization: Bearer <KNOCK_API_KEY>' \
--data '{
  "name": "<tenant-name>",
}'
```

#### Create an object

> In Knock, [objects](https://docs.knock.app/concepts/objects) are flexible abstractions, and with SlackKit they are used to store channels and act as the recipient of your workflows. You can't create a new object from the dashboard, so you can use this cURL command to create an object by replacing the values for `object-collection`, `object-id`, `KNOCK_API_KEY`, and `object-name`:

Reviewer: ```suggestion
In SlackKit, [objects](https://docs.knock.app/concepts/objects) are used to store channels and act as the recipient of your workflows. You can't create a new object from the dashboard, so you can use this cURL command to create an object by replacing the values for `object-collection`, `object-id`, `KNOCK_API_KEY`, and `object-name`:
```

```
import "@knocklabs/react/dist/index.css";
import jwt from "jsonwebtoken";

import Providers from "./components/providers";
import "./global.css";
import { getAppDetails } from "./lib/app-details";

const { userId, tenant, collection, objectId } = getAppDetails();

const currentTime = Math.floor(Date.now() / 1000);
const expireInSeconds = 60 * 60;
const signingKey = process.env.KNOCK_SIGNING_KEY!;

const userToken = signingKey
  ? jwt.sign(
      {
        sub: userId,
        iat: currentTime,
        exp: currentTime + expireInSeconds,
        grants: {
          [`https://api.knock.app/v1/objects/$tenants/${tenant}`]: {
            "slack/channels_read": [{}],
          },
          [`https://api.knock.app/v1/objects/${collection}/${objectId}`]: {
            "channel_data/read": [{}],
            "channel_data/write": [{}],
          },
        },
      },
      signingKey,
      {
        algorithm: "RS256",
      },
    )
  : "secretOrPrivateKey";

> 

Reviewer: we probably want to replace this with chris' new SDK right? https://github.com/knocklabs/knock-node/pull/56

Author: Yeah, looks like that's still open, so I'll loop back pre-release and update this

```
"use client";

import { SlackChannelCombobox } from "@knocklabs/react";

export default function SlackChannelWrapper({
  collection,
  objectId,
  className,
}: {
  collection: string;
  objectId: string;
  className: string;
}) {
  const slackChannelsRecipientObject = {
    objectId: objectId,
    collection: collection,
  };
  return (
    <div className={className}>
      <SlackChannelCombobox slackChannelsRecipientObject={slackChannelsRecipientObject} />
    </div>
  );
}
```

As you select channels using the combobox input, the component will update the channel data for that object to store the Slack channel ids.

![choose Slack channels](./images/choose-slack-channel.png)

**Note: if you want to use a private Slack channel, you need to make sure that the Slack Bot has been invited to that channel.**

At this point, Knock has all of the information it needs to send Slack messages to your selected channels. Click `Next` to examine the channel data for your object and tenant.

## Examine channel data

> In this step, you're not required to take any action, but this page uses the Node SDK to pull channel data for your object recipient and tenant. Hopefully, this gives you an idea of how Knock is storing the data and these two abstractions interact.

Reviewer: ```suggestion
In this step, you're not required to take any action, but this page uses the Node SDK to pull channel data for your object recipient and tenant. This will give you an idea of how Knock is storing the data and these two abstractions interact.
```

This data is pulled dynamically, so if you make different channel selections or disconnect and reconnect your Slack auth, the value for `access_token` and `connections.channelIds` will change as well.

If you click `Next` you can test your workflow end-to-end.

## Trigger your workflow

In this step, you can see a code sample using the Node SDK of how you'd trigger your workflow. This is actually the code that runs in a server action when you submit the form on this page. Go ahead and submit a form to send a Slack message into your designated channels:

![trigger a workflow](./images/trigger-workflow.png)

## Wrapping up

> We know this is a long tutorial, but just imagine how long it would be if you had to do all of this without SlackKit components and APIs. If you need help implementing any of the code in this example app, consider joining our [Slack community](https://knock.app/join-slack). We've got over 500 other developers in there learning how to improve their notification experiences with Knock.

Reviewer: I'm not sure we need to note this here but i'm not the marketing person so defer to y'all 🤓 

```suggestion
If you need help implementing any of the code in this example app, consider joining our [Slack community](https://knock.app/join-slack). We've got over 500 other developers in there learning how to improve their notification experiences with Knock.
```

Author: This might be my best line of copywriting, but I can save it for the blog post version
